File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/IconListAdapter.java
package de.szalkowski.activitylauncher;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.BaseAdapter;
import android.widget.ImageView;

import java.util.List;
import java.util.TreeSet;

public class IconListAdapter extends BaseAdapter {
    private String[] icons;
    private PackageManager pm;
    private Context context;

    IconListAdapter(Context context) {
        this.context = context;
        this.pm = context.getPackageManager();
    }

    static Drawable getIcon(String icon_resource_string, PackageManager pm) {
        try {
            String pack = icon_resource_string.substring(0, icon_resource_string.indexOf(':'));
            String type = icon_resource_string.substring(icon_resource_string.indexOf(':') + 1, icon_resource_string.indexOf('/'));
            String name = icon_resource_string.substring(icon_resource_string.indexOf('/') + 1);
            Resources res = pm.getResourcesForApplication(pack);
            return res.getDrawable(res.getIdentifier(name, type, pack));
        } catch (Exception e) {
            return pm.getDefaultActivityIcon();
        }

    }

    void resolve(IconListAsyncProvider.Updater updater) {
        TreeSet<String> icons = new TreeSet<>();
        List<PackageInfo> all_packages = this.pm.getInstalledPackages(0);
        updater.updateMax(all_packages.size());
        updater.update(0);

        PackageManagerCache cache = PackageManagerCache.getPackageManagerCache(this.pm);

        for (int i = 0; i < all_packages.size(); ++i) {
            updater.update(i + 1);

            PackageInfo pack = all_packages.get(i);
            try {
                MyPackageInfo mypack = cache.getPackageInfo(pack.packageName);

                for (int j = 0; j < mypack.getActivitiesCount(); ++j) {
                    String icon_resource_name = mypack.getActivity(j).getIconResouceName();
                    if (icon_resource_name != null) {
                        icons.add(icon_resource_name);
                    }
                }
            } catch (NameNotFoundException ignored) {
            }
        }

        this.icons = new String[icons.size()];
        this.icons = icons.toArray(this.icons);
    }

    @Override
    public int getCount() {
        return icons.length;
    }

    @Override
    public Object getItem(int position) {
        return icons[position];
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ImageView view = new ImageView(this.context);
        AbsListView.LayoutParams layout = new AbsListView.LayoutParams(50, 50);
        view.setLayoutParams(layout);
        String icon_resource_string = this.icons[position];
        view.setImageDrawable(IconListAdapter.getIcon(icon_resource_string, this.pm));
        return view;
    }
}

File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/IconListAdapter.kt
package de.szalkowski.activitylauncher

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.PackageManager.NameNotFoundException
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.BaseAdapter
import android.widget.ImageView
import java.util.TreeSet

class IconListAdapter internal constructor(private val context: Context) : BaseAdapter() {
    private var icons: Array<String>? = null
    private val pm: PackageManager

    init {
        this.pm = context.packageManager
    }

    internal fun resolve(updater: AsyncProvider<IconListAdapter>.Updater) {
        val all_packages = this.pm.getInstalledPackages(0)
        updater.updateMax(all_packages.size)
        updater.update(0)

        val cache = PackageManagerCache.getPackageManagerCache(this.pm)

        this.icons = all_packages.mapIndexed {
            i, pack ->

            updater.update(i + 1)
            try {
                val info = cache.getPackageInfo(pack.packageName)
                (0 until info!!.activitiesCount).mapNotNull { info.getActivity(it).iconResouceName }
            } catch (ignored: NameNotFoundException) {
                emptyList<String>()
            }
        }.flatten().toHashSet().toTypedArray()
    }

    override fun getCount(): Int {
        return icons!!.size
    }

    override fun getItem(position: Int): Any {
        return icons!![position]
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getView(position: Int, convertView: View, parent: ViewGroup): View {
        val view = ImageView(this.context)
        val layout = AbsListView.LayoutParams(50, 50)
        view.layoutParams = layout
        val icon_resource_string = this.icons!![position]
        view.setImageDrawable(IconListAdapter.getIcon(icon_resource_string, this.pm))
        return view
    }

    companion object {

        internal fun getIcon(icon_resource_string: String, pm: PackageManager): Drawable {
            try {
                val pack = icon_resource_string.substring(0, icon_resource_string.indexOf(':'))
                val type = icon_resource_string.substring(icon_resource_string.indexOf(':') + 1, icon_resource_string.indexOf('/'))
                val name = icon_resource_string.substring(icon_resource_string.indexOf('/') + 1)
                val res = pm.getResourcesForApplication(pack)
                return res.getDrawable(res.getIdentifier(name, type, pack))
            } catch (e: Exception) {
                return pm.defaultActivityIcon
            }

        }
    }
}
